Add simple support for STEM files

acolombier opened the pull request:
This feature allows loading STEM file, extract the stem data and ensure the correctness of the file.

The SoundSource will provide the main mix (the first stream by convention) to the audio engine and the stem info data to the track object.

This PR depends on #13042

acolombier committed:
feat: add support for STEM

The soundsource will ensure the file is a valid STEM file and load
manifest. It will however only expose the main mix to the sound engine

daschuer commented:
Can you rebase this on main, to get rid of the first commit?

JoergAtGithub commented on src/sources/soundsource.cpp:
```diff
@@ -32,7 +32,7 @@ QString SoundSource::getTypeFromUrl(const QUrl& url) {
 //static
 QString SoundSource::getTypeFromFile(const QFileInfo& fileInfo) {
-    const QString fileSuffix = fileInfo.suffix().toLower().trimmed();
+    const QString fileSuffix = fileInfo.completeSuffix().toLower().trimmed();
```
This would change the file type recognition for files names like `artistname.songtitle.flac`
I'm not sure if this case is relevant, but `fileinfo.filename().toLower().trimmed().endsWith(".stem.mp4")` below would be the safe way to implement this.

JoergAtGithub commented on src/track/steminfoimporter.cpp:
```diff
@@ -0,0 +1,139 @@
+/// @return the size of the box data if found, -1 otherwise
+uint32_t seekTillAtom(QIODevice& reader,
+        const uint32_t path[],
+        uint32_t box_size = 0,
```
```suggestion
        uint32_t boxSize = 0,
```

JoergAtGithub commented on src/track/steminfoimporter.cpp:
```diff
@@ -0,0 +1,139 @@
+
+// static
+bool StemInfoImporter::isStemFile(
+        const QString& aFileName) {
```
```suggestion
        const QString& fileName) {
```

JoergAtGithub commented on src/track/steminfoimporter.cpp:
```diff
@@ -0,0 +1,139 @@
+
+    // Extract DSP information
+    // TODO(XXX) DSP only contains limit and a compressor, which aren't
+    // supported by Mixxx yet. parse and implement when supported
```
```suggestion
    // used by Mixxx yet. Parse and implement when supported
```

acolombier replied:
I'm afraid this won't mean what I tried to say here. I have reworded it differently now, let me know if that makes more sense.

JoergAtGithub replied:
Comment is clear now!

JoergAtGithub commented on src/track/steminfoimporter.h:
```diff
@@ -0,0 +1,18 @@
+
+namespace mixxx {
+
+/// Importer class for StemInfo objects that can correct timing offsets when the
+/// signal info (channel number, sample rate, bitrate) is known.
```
This comment seems not to fit here.

JoergAtGithub commented on src/sources/soundsourcestem.h:
```diff
@@ -0,0 +1,69 @@
+class SoundSourceSingleSTEM : public SoundSourceFFmpeg {
+  public:
+    // streamIdx is the FFmpeg stream id, which may different than stemIdx + 1
+    // because STEm may contain other non audio stream
```
```suggestion
    // because STEM may contain other non audio stream
```

JoergAtGithub reviewed (commented):
Added some minor remarks

acolombier committed:
Revert breaking change on soundsource and improve docstrings

JoergAtGithub commented on CMakeLists.txt:
```diff
@@ -3470,6 +3444,51 @@ if(FFMPEG)
+  target_sources(mixxx-lib PRIVATE
+    src/sources/soundsourcestem.cpp
+    src/track/steminfoimporter.cpp
+    src/track/steminfo.cpp # TODO precompiled header?
```
Yes, please add track/steminfo.h to the list of precompiled headers. In the follow-up PRs it's included at various places an makes therefore sense.

JoergAtGithub replied:
Please add the header to the precompiled header list

acolombier replied:
I haven't figured out how to make that conditionally yet and will add once I have a solution. Happy to do right away if you have a recommended way.

acolombier replied:
Looks like `target_precompile_headers` adds headers and don't override the list as I thought. Added now

JoergAtGithub commented on src/sources/soundsourcestem.cpp:
```diff
@@ -0,0 +1,468 @@
+                        m_buffer.data(),
+                        stemSampleLength));
+        m_pStereoStreams[streamIdx]->readSampleFrames(currentStemFrame);
+        // TODO(XXX): currently, stem samples are interleaved and packed next to each other as such:
+        //    1L1R1L1R1L1R...2L2R2L2R2L2R2L2R......3L3R3L3R3L3R3L3R......4L4R4L4R4L4R4L4R....
+        //    Can FFmpeg decode as without having to use a decoder per channel?
+        //    1LLLLLLLLLLLLLL....1RRRRRRRRR...2LLLLLLL...?
+
+        // Change the sample layout to interleave all channels together
+        for (SINT i = 0; i < stemSampleLength / 2; i++) {
+            pBuffer[2 * stemCount * i + 2 * streamIdx] = m_buffer[2 * i];
+            pBuffer[2 * stemCount * i + 2 * streamIdx + 1] = m_buffer[2 * i + 1];
+        }
```
I assume that this memory structure is to be optimized across all modules in follow-up PR?

acolombier replied:
I guess by optimisation, you mean have FFmpeg take care of the layout when decoding? I don't have any immediate plan. I had a quick look and I didn't find any simple solutions for this as each stream is technically independent, handled by a dedicated FFmpeg context. Do you have any suggestion or tips on how to do so? 
FYI, I took some stats to see if that for-loop/re-layout was a bottleneck, and it is about 1/100th of the time used by RubberBand so probably not a problem yet.

JoergAtGithub replied:
The rubberband performance is a huge problem - was already before, but stems multiplies this. On my rather fast Windows laptop, I can only use a singles stems deck with Rubberband R2 and R3 becomes unusable with stems.
I could imagine, that one major reason for this is, that the data are not alligned in the memory, to allow the compiler to use efficient SIMD instructions. Maybe this can only be changed in Rubberband itself, but maybe already the memory alignment in our engine might influence this.
Therefore I think, that analysis and overall optimization of the engines memory structures might be necessary.

acolombier replied:
> but maybe already the memory alignment in our engine might influence this

This part is already aligned, but I assume you me the [channel buffer](https://github.com/acolombier/mixxx/blob/b8c19a0c2ba6b77971d2e15cf84faa258610e018/src/engine/bufferscalers/enginebufferscalerubberband.cpp#L166)

Are we sure RubberBand would use SIMD? If yes, I can try to scrape the individual `SampleBuffer` and use a a contiguous mem alloc, instead of individual ones

acolombier replied:
(Note that I had plenty of `ScopedTimer` all across which allow me to confirm that 99% of the CPU time is due to RubberBand::retrieve, which I removed now)

acolombier replied:
I did the experiment with a contiguous global buffer and there was no changes.
Just FYI, I've put back the `ScopedTimer`  and here is the result
```
Stat("EngineBufferScaleRubberBand::scaleBuffer","count=462,sum=2.82884e+09ns,average=6.12302e+06ns,min=2.77938e+06ns,max=1.97878e+07ns,variance=3.76359e+12ns^2,stddev=1.94e+06ns")
Stat("EngineDeck::processStem","count=535,sum=2.88642e+09ns,average=5.39517e+06ns,min=89170ns,max=2.01299e+07ns,variance=7.62824e+12ns^2,stddev=2.76193e+06ns")
Stat("RubberBand::process","count=1839,sum=2.81529e+09ns,average=1.53088e+06ns,min=7860ns,max=4.10194e+06ns,variance=8.39907e+10ns^2,stddev=289812ns")
Stat("RubberBand::retrieve","count=2299,sum=2.65024e+06ns,average=1152.78ns,min=105ns,max=16231ns,variance=1.78589e+06ns^2,stddev=1336.37ns")
```
(Just the sum)
```
EngineDeck::processStem                  2.88642e+09ns
EngineBufferScaleRubberBand::scaleBuffer 2.82884e+09ns
RubberBand::process                      2.81529e+09ns
RubberBand::retrieve                     2.65024e+06ns
```
It was `RubberBand::process` btw, not `RubberBand::retrieve`

I guess another experiment would be to aligned the buffer size with the [sample count](https://github.com/mixxxdj/mixxx/blob/b8c19a0c2ba6b77971d2e15cf84faa258610e018/src/engine/bufferscalers/enginebufferscalerubberband.cpp#L261), I'll try that tomorrow.

JoergAtGithub replied:
Great, that you did the analysis already! It would be even greater, to store such results in a document.
A helpul tool to understand such performance issues is callgrind. I used it once (for a much smaller single thread project), and could clearly see, where the CPU cycles were burned. And I could see, where the compiler picked SIMD implementations of std library functions, and where not.

acolombier committed:
Add support for .m4a extension variation

acolombier commented:
I have added the support for `.stem.m4a` extension as well. Note that it isn't strictly mentioned in the spec (or maybe I missed it?) but it appears that Traktor supports it as well, and so project like [stemgen](https://github.com/axeldelafosse/stemgen) seems to generate this extension by default.

daschuer commented on src/sources/soundsource.cpp:
```diff
@@ -52,6 +53,18 @@ QString SoundSource::getTypeFromFile(const QFileInfo& fileInfo) {
+    if (normalisedFilename.endsWith(QLatin1String(".stem.m4a"))) {
+        // STEM suffix may also extends the ".m4a" suffix for the same reason as above
+        return QLatin1String("stem.m4a");
+    }
```
We normally prioritize the mime type over the file extension. I think mp4 and m4a both have "audio/mp4".
So I think we need to look up the mime type first and than check if the file name is "*stem.*". This will finally allow any random file extension which has a stem file name. Or we may be more strict and limit it to "stem.m4a" and "stem.mp4".
Optimal would be a magic number to look up. But probably not worth the effort.

daschuer replied:
At which offset are the atoms located? Is the lookup fast? Maybe we can do this.

daschuer replied:
Cool, than let's do it. 
Is it correct to do it for all audio/mp4 mime types?

acolombier replied:
Sounds good to me.

daschuer commented on src/sources/soundsourceproxy.cpp:
```diff
@@ -177,6 +180,13 @@ QList<QMimeType> mimeTypesForFileType(const QString& fileType) {
+        // *.stem.mp4 and *.stem.m4a suffix do not have a specific MIME type associated
+        // with them, and simply fall back to MP4. To prevent conflicts with the
+        // MP4 decoder able to decode arbitrary MP4 file (such as video,
+        // extracting just the audio track), we don't return any MIME here
```
I think they have. At least libmagic (via Qt) will return one, I guess

daschuer replied:
Ok, so please adjust the comments accordingly. 
Maybe this will receive some refactoring when implementing the stem lookup. 

Another (crazy) thought:
I can imagine to put two or more aac streams into an MP4. How to deal with them, play only the first stream like we do now or treat them like stem files?

daschuer replied:
OK, sounds reasonable.

daschuer commented on src/sources/soundsourcestem.cpp:
```diff
@@ -0,0 +1,468 @@
+const QString SoundSourceProviderSTEM::kDisplayName = QStringLiteral("STEM with FFmpeg");
+
+QStringList SoundSourceProviderSTEM::getSupportedFileTypes() const {
+    return {"stem.mp4", "stem.m4a"};
```
Is one with video? Or can we normalize them to only one.

acolombier replied:
NI suggest using `stem.mp4` and provide stem samples with the `stem.mp4`, however a few tools online (such as stemgen) generate `stem.m4a` files. Traktor supports both `mp4` and `m4a` extension.

daschuer replied:
OK

daschuer commented on src/sources/soundsourcestem.cpp:
```diff
@@ -0,0 +1,468 @@
+        : SoundSourceFFmpeg(url), m_streamIdx(streamIdx) {
+}
+
+SoundSource::OpenResult SoundSourceSingleSTEM::tryOpen(
```
Is there an early check if this is really a stem file? If not, it should fail in the first attempt. Maybe also later and the normal m4a decoder can take over.

acolombier replied:
There is a check that ensure the audio tracks have matching spec. Is there other checks you would like to see?

daschuer replied:
I have no specific demand. Just keep in mind that this tryOpen() is called according to the priority of a certain format.
In other sound sources there is a Strict and a Permissive mode. I have no idea if this here reasonable as well.

dszakallas replied:
wouldn't it be possible to try to create a stem soundsource for any mp4, and if we don't find the metadata, or we face other validation error, continue with the generic mp4 reader?

acolombier replied:
We could do that but that feels wrong and biased to assume that MP4 == STEM IMO. What is the the team opinion?

daschuer replied:
We have a probing facility already. 
If the file is a stem.mp4 it is probed here first. 
In case tryOpen() fails, libfaad should take over.

daschuer reviewed (changes requested):
This is the first batch of comments.

daschuer commented on src/track/steminfoimporter.cpp:
```diff
@@ -0,0 +1,144 @@
+} // anonymous namespace
+
+// static
+bool StemInfoImporter::isStemFile(
```
Here we need apply the mime type logic. And probably the checking for easy "stem" properties in addition.
Alternative  this may look for these properties only and we can call the function isMp4AStem() or such.  
This raises the general question if this Class can be only used with MP4s or if we need to protect every invocation to be called with other files.

acolombier replied:
> Here we need apply the mime type logic

AFAIK, mime type isn't available in the Track object. Reading the file (to detect the MIME using `QMimeDatabase::mimeTypeForFile` will likely have effect on performance. of every single file being imported, which is likely not negligible when importing a large library. 

> This raises the general question if this Class can be only used with MP4s

I'm not sure why we would want to support another file type. The STEM atom is part of the NI standard, meaning it needs to be in a mp4 file. If we want to implement other STEM standards, there is no guarantee that there will use a STEM atom, as part of the MPEG standard. My suggestion would be to keep the implantation to what we know and support currently (NI standard, mp4) and cross the bridge to support other standards when we decide to do so.

dszakallas replied:
I think that that detection based on `stem.mp4` extension is overly restrictive. At the extreme any iso media container could be feature tested for `moov.udta.stem`, but at the least we should be able to detect stems in arbitrary `.mp4` containers.

acolombier replied:
Again, I was simply implementing the specification (which I believe @dszakallas also had access to). Happy to go beyond what it defines.

JoergAtGithub replied:
We should not expand this PR by general MP4 support. Scope of this PR is the STEMs format and nothing beyond!
Traktor Pro seems to have another detection logic than just checking for .stems.mp4 extension, because it seems to work also with the extension .stems.m4a (or they just check for one of both file extensions).
Performance for file type detection is important in my opinion. What would it cost to probe for moov.udta.stem ?

dszakallas replied:
We already support mp4 some way, as I was able to drop mp4 tracks into mixxx and play them. And the specification says that the container should be an ISOBMFF with `.mp4` as extension. `.stem.mp4` is "preferred", not required. So, if we want to follow the spec we should try to detect stems in any mp4.

acolombier replied:
> What would it cost to probe for moov.udta.stem ?

That's hard to assess as the atom offset isn't static and require to traverse boxes in the file. 

> extension. .stem.mp4 is "preferred", not required

That's true but in the meantime, Mixxx aims to be vendor-agonistic IMO, so the flexibility that Tracktor Pro gives to arrange NI shouldn't drive Mixxx implementation. Anyway, last commit adds support for `.mp4`, but happy to remove depending where the consensus goes.

daschuer commented on src/track/steminfoimporter.cpp:
```diff
@@ -0,0 +1,144 @@
+/// search in
+/// @param pathIdx The level of the tree currently search (index of path)
+/// @return the size of the box data if found, -1 otherwise
+uint32_t seekTillAtom(QIODevice& reader,
```
We avoid non const references in Mixxx. Please pass the reader as a pointer.

daschuer commented on src/track/steminfoimporter.cpp:
```diff
@@ -0,0 +1,144 @@
+    auto json = jsonData.object();
+
+    // Check the manifest version
+    auto version = json.value("version").toInteger(-1);
```
Can you add here a typical JSON as a source code comment as a reference? 
Maybe also a link to the specification.

JoergAtGithub replied:
Maybe we could add the JSON from src/test/stems/test.stem.mp4 as .json file next to it?

acolombier replied:
Done

daschuer commented on src/track/steminfoimporter.cpp:
```diff
@@ -0,0 +1,144 @@
+        auto name = stem.value("name").toString();
+        if (!color.isValid() || name.isEmpty()) {
+            kLogger.debug() << "Unexpected or missing stem name or attribute in STEM manifest";
+            return {};
```
I think we should be less critical and pick reasonable defaults. A stem file should be play-able, with a minimum of information that is required to play it.

acolombier replied:
"Stem #<n>" and Red/Green/Blue/Yellow?

acolombier replied:
Not entirely pure, but some red, green and blue are used by default for waveform as mentioned in my [3-band waveform PR](https://github.com/mixxxdj/mixxx/pull/13075)

JoergAtGithub replied:
Let's use the ones from stemgen, as the practical use case is stem splitting with stemgen and than use the generated files in Mixxx.

daschuer commented on src/track/steminfoimporter.cpp:
```diff
@@ -0,0 +1,144 @@
+        ATOM_TYPE("moov"),
+        ATOM_TYPE("udta"),
+        ATOM_TYPE("stem"),
+        0 // This indicate end of the path
```
What happens if moov and udta are missing? I think it would be more robust to just allow any atom that is before stem.

daschuer replied:
Ah, so I have probably misread the code. 
Do the atoms build a folder structure? 
In that, case everything is fine.

acolombier replied:
Yes, they are organise as a tree structure, with multiple root (`moov` being one of them)

daschuer commented on src/track/steminfoimporter.cpp:
```diff
@@ -0,0 +1,144 @@
+        if (ATOM_TYPE(buffer + 4) == path[pathIdx]) {
+            return seekTillAtom(reader, path, ATOM_TYPE(buffer) - kAtomHeaderSize, pathIdx + 1);
+        }
+        reader.skip(ATOM_TYPE(buffer) - kAtomHeaderSize);
```
ATOM_TYPE should not be used to also return the size. 
It would be more easy to read to cast to a struct that contains type and size.

JoergAtGithub commented on src/track/taglib/trackmetadata_file.cpp:
```diff
@@ -48,12 +48,16 @@ namespace taglib {
         const QString& fileName) {
-    DEBUG_ASSERT(!fileName.isEmpty());
-    const QString fileSuffix(fileName.section(QChar('.'), -1).toLower().trimmed());
+    QString normalisedFileName = fileName.toLower().trimmed();
```
There is no dedicated mime-type assigned for the STEM format: https://www.iana.org/assignments/media-types/media-types.xhtml#audio
There are only two options, either using the file extension, or reading the content of the file and analysing it (which is time consumig).

daschuer replied:
In this case this is a taglib function only used to check for mp3. There should be no need to deal with stems here.

acolombier replied:
> this is a taglib function only used to check for mp3

I'm not sure this is. This function is used by TagLib to decide which taglib reader shoud be used to extract metadata.

daschuer replied:
Ok, so we have found an issue with our mp3 with wav case, or for files without extension?

acolombier replied:
Yes, but this should only be impacting metadata tho, you should still be able to load and read the file.

acolombier replied:
@daschuer do you me to create an issue some taglib also relies on file MIME instead of extension?

acolombier commented on src/track/taglib/trackmetadata_file.cpp:
```diff
@@ -48,12 +48,16 @@ namespace taglib {
         return FileType::MP3;
     }
-    if ((QStringLiteral("m4a") == fileSuffix) || (QStringLiteral("m4v") == fileSuffix)) {
+    if ((QStringLiteral("m4a") == fileSuffix) ||
+            (QStringLiteral("m4v") == fileSuffix) ||
+            normalisedFileName.endsWith("stem.mp4") ||
+            normalisedFileName.endsWith("stem.m4a")) {
```
```suggestion
    if ((QStringLiteral("m4a") == fileSuffix) ||
            (QStringLiteral("m4v") == fileSuffix) ||
            normalisedFileName.endsWith("stem.mp4")) {
```

Or do we want to support `mp4` entirely?

acolombier commented on src/track/track.h:
```diff
@@ -495,6 +511,17 @@ class Track : public QObject {
+#ifdef __STEM__
+    /// Sets stem info and returns a boolean to indicate if stems were updated.
+    /// Only supposed to be called while the caller guards this a lock.
+    bool setStemPointsWhileLocked(const QList<StemInfo>& stemInfo);
```
Probably a mistake when fixing typos, will rename to `StemInfo`

dszakallas commented on src/track/steminfoimporter.cpp:
```diff
@@ -0,0 +1,144 @@
+        reader.read(buffer, kAtomHeaderSize);
+        byteRead += boxSize;
+        if (ATOM_TYPE(buffer + 4) == path[pathIdx]) {
+            return seekTillAtom(reader, path, ATOM_TYPE(buffer) - kAtomHeaderSize, pathIdx + 1);
```
There's also the possibility of 64 bit box; in case `atomsize==1` there's an additional [8 bytes encoding the size](https://github.com/DigiDNA/ISOBMFF/blob/main/ISOBMFF/source/ContainerBox.cpp#L101).

dszakallas commented on src/track/steminfoimporter.cpp:
```diff
@@ -0,0 +1,144 @@
+
+    // Check the manifest version
+    auto version = json.value("version").toInteger(-1);
+    if (version <= 0) {
```
I'd rather simplify all this by checking for `version==1` and bailing out on anything else.

acolombier replied:
I'm confused. In other comments, we aim to make the implementation be permissive and allow flexibility. 
Happy to change the debug at `108` to warning if we want it to be visible with the default log level, but it goes against the other efforts to hard fail here for no reason.

dszakallas replied:
this is mostly theoretical as there was no update to the this standard since released nearly 10 years ago, but since this is a single number, an increment likely means a breaking change. so I think it would be cleaner to outright reject something that we have no guarantee of parsing correctly

acolombier replied:
> I think it would be cleaner to outright reject something that we have no guarantee of parsing correctly

That statement can be applied on many of the other requested changes. Again, I think it's best to keep and trying best effort on continuing loading. There are more tests coming later to validate the data (channel layout, matching sample rate, ...) so if this is indeed a breaking change, then it will abort later one, when it gets indeed stuck. At that stage, it's best to carry on IMO. 
As a middle ground, I will update the subsequent `debug` log to a `warning` so it gets displayed by default.

dszakallas commented on src/track/steminfoimporter.cpp:
```diff
@@ -0,0 +1,144 @@
+
+    // Extract DSP information
+    // TODO(XXX) DSP only contains a limiter and a compressor effect, which
+    // Mixxx doesn't have yet. Parse and implement when supported
```
nit: maybe worth printing a warning if we detected enabled compressor/ limiter the fact that it will be ignored.

acolombier replied:
Do you mean a modal popup? If you mean log messages, I'm not sure what would be the value of it, but happy to put a debug a message to reduce noise.

JoergAtGithub replied:
This information would not be helpful for a normal user. A warning in the log should be enough.

acolombier replied:
I'll put a debug message

acolombier commented on CMakeLists.txt:
```diff
@@ -3577,8 +3596,8 @@ if(OPUS)
   target_compile_definitions(mixxx-lib PUBLIC __OPUS__)
-  target_link_libraries(mixxx-lib PRIVATE OpusFile::OpusFile)
-  target_link_libraries(mixxx-lib PRIVATE Opus::Opus)
+  target_link_libraries(mixxx-lib PRIVATE OpusFile::OpusFile Opus::Opus)
```
It only mean that the tests won't include Opus library, and test case will be omitted (eg soundsourceproxy). So while it impacts 2.4, it probably also have no impact. What's your suggestion?

daschuer replied:
Ok, I understand. So let's keep it as it is.

acolombier committed:
Detect stem file in arbitrary .mp4 file

JoergAtGithub commented on src/track/steminfoimporter.cpp:
```diff
@@ -1,31 +1,57 @@
 
 const mixxx::Logger kLogger("StemInfoImporter");
 constexpr int kSupportedStemVersion = 1;
 const QStringList kStemFileExtensions = {".stem.mp4", ".stem.m4a"};
```
As far as I learned, there are three cases:

- Anything that ends with `.mp4` is allowed
- `.stem.mp4` is recommended
- `.stem.m4a` is not specified, but is used in practice (Stemgen writes it and Traktor Pro reads it)

acolombier replied:
Yes, that's correct and is already how it work. STEM extension are only defines here as fallback here in case the MIME detection fails from file content. Should I put a comment to explicitly mention that these extension are only used as fallback?

JoergAtGithub replied:
Yes, a commet would help here!

JoergAtGithub commented on src/track/steminfoimporter.cpp:
```diff
@@ -0,0 +1,191 @@
+
+constexpr uint32_t kAtomHeaderSize = sizeof(MP4BoxHeader);
+
+/// @brief Deserialise a MP$ atom/box header
```
```suggestion
/// @brief Deserialise a MP4 atom/box header
```

JoergAtGithub commented on src/sources/soundsource.cpp:
```diff
@@ -52,9 +53,18 @@ QString SoundSource::getTypeFromFile(const QFileInfo& fileInfo) {
     QMimeType mimeType = QMimeDatabase().mimeTypeForFile(
             fileInfo, QMimeDatabase::MatchContent);
+    if (
+            // STEM files will be detected as normal MP4 se we check if the file
```
```suggestion
            // STEM files will be detected as normal MP4, so we check if the file
```

daschuer commented:
Just did a test: 
* Remove suffix form an mp3 file. 
* Library scanner ignores it 
* Can't be dropped to decks. 

* Changed suffix form mp3 to wav.
* Library scanner finds the file type is correctly shown as mp3 
* Metadata is not imported 
* Track can be played 
* Track can be dropped to the deck

acolombier committed:
Fix typo and add comments

acolombier commented:
Thanks for these tests @daschuer 

I can confirm that this is now the same behaviour with STEM files:

With `wav`

- Changed suffix form stem.mp4 to wav.
- Library scanner finds the file type is correctly shown as stem.mp4
- Metadata is not imported
- Track can be played
- Track can be dropped to the deck

With `m4a`

- Changed suffix form stem.mp4 to m4a.
- Library scanner finds the file type is correctly shown as stem.mp4
- Metadata **is** imported
- Track can be played
- Track can be dropped to the deck

acolombier commented:
Note that, as faired, I have noticed a performance impact on probing the STEM box in each MP4 file, especially when running the tests. This is slightly worrying as I have a pretty fast CPU and NVME, so I assume that performance impact will be noticeable.

JoergAtGithub commented:
> Note that, as faired, I have noticed a performance impact on probing the STEM box in each MP4 file, especially when running the tests. This is slightly worrying as I have a pretty fast CPU and NVME, so I assume that performance impact will be noticeable.

Does it only consumes that much time, if it is a STEM file, or also for normal MP4 audio files?

acolombier commented:
Hard to say - I suspect something is wrong in the way the file type is being detected, leading to the `SoundSourceProxy` test to timeout. I will do further investigation, and come back with further info.

acolombier commented:
After running the `SoundSourceProxyTest` through callgrind, it looks like the probing of the stem atom was too greedy and could stop recursion early on. There is still a slight performance impact but that should be acceptable now.

acolombier commented on src/track/track.cpp:
```diff
@@ -1296,6 +1296,21 @@ bool Track::importPendingCueInfosWhileLocked() {
 }
 
+#ifdef __STEM__
+bool Track::setStemPointsWhileLocked(const QList<StemInfo>& stemInfos) {
```
I have updated the setter, but noticed that some others in this file also relies on copy by ref (eg. setCuePointsWhileLocked). Should I updated them?

JoergAtGithub commented on CMakeLists.txt:
```diff
@@ -3470,6 +3444,54 @@ if(FFMPEG)
+    src/test/stemtest.cpp
+    src/test/steminfotest.cpp
+  )
+  target_precompile_headers(mixxx-lib PUBLIC
+    src/track/steminfo.h
+  )
```
I guess only one `target_precompile_headers` statement per target is possible. Try to add the file as entry after line 1334 encapsulated in `if (STEM)` `endif()`

acolombier replied:
Looks like Windows is still not happy. Not sure why only `soundsource.cpp` has unresolved symbol. Or does MSVC compiler only display the first object missing symbols?

acolombier committed:
Fix cmake precompiled header not building on Windows

JoergAtGithub reviewed (approved):
I compiled this locally on Windows11 and it compiles fast and tests pass. All my change requests are addressed!

daschuer commented on src/sources/soundsourcestem.h:
```diff
@@ -0,0 +1,69 @@
+};
+
+/// @brief Handle a stem file, composed of multiple audio channel. Can open in
+/// stereo or in stem (4 x stereo)
```
Can the the user of the sound source decide? Probably not. Can you refine the comment?

We have the issue that pure mp4 files are opened with libfaad which is not offset compatible. This means we will have the issue that before merging this PR *.stem.mp4 have been decoded with a different offset and the analysis is no longer consistent. We need to deal with it somehow. Suggestions?

JoergAtGithub replied:
Since we can (in future) load a STEM file as well into a deck (all 4 stems seperated) as into a sampler/preview deck (either mixed stereo OR a selected single stem stereo track), STEM files need always to be analyzed with the same offset, independend if we use the mixed stereo, or the stem tracks.
While non-STEM .mp4 are completly independent.

I think we are safe, if we can safely distinguish STEM files from non-STEM .mp4 files and read STEM files always using FFMPEG.

acolombier replied:
@JoergAtGithub is correct - when you open a SoundSource, you pass the `OpenParams` which allow to request a maximum number of supported channel (since Mixxx uses the StereoProxy in case the channel count is less than requested). This allows decks which must not use STEM for performance or usability reason to use the soundsource as a strict stereo (samples, preview, but also analyser currently, till the PR which introduce STEM analysis)

acolombier committed:
Update docstring for STEM soundsource

Co-authored-by: Daniel Schürmann <daschuer@mixxx.org>

dszakallas commented on src/sources/soundsourcestem.h:
```diff
@@ -0,0 +1,71 @@
+
+namespace mixxx {
+
+/// @brief Handle a single stem stem embedded in a stem file
```
```suggestion
/// @brief Handle a single stem embedded in a stem file
```

acolombier committed:
Update src/sources/soundsourcestem.h

Co-authored-by: Dávid Szakállas <5807322+dszakallas@users.noreply.github.com>

acolombier commented:
@daschuer friendly ping in case you missed the notification

dszakallas commented on src/track/steminfoimporter.cpp:
```diff
@@ -0,0 +1,215 @@
+                           "this isn't supported by Mixxx yet.";
+    }
+
+    file.close();
```
code style: isn't it a problem that this is not called on earlier exit paths of this func?

acolombier replied:
[QFileDevice](https://doc.qt.io/qt-6/qfiledevice.html#dtor.QFileDevice) (which QFile inherit of) will ensure it closes it if necessary on destruction

daschuer reviewed (approved):
LGTM, Thank you. 

Sorry for the delay. I have underestimated the time for branching out 2.5 compared to my free time this week.
